Language/Haskell/GhcMod/ImportedFrom.hs

```diff
@@ -0,0 +1,785 @@
+-- Copyright (C) 2013-2016 Carlo Hamalainen <carlo ÄT carlo-hamalainen DOT net>
+--
+-- This program is free software: you can redistribute it and/or modify
+-- it under the terms of the GNU Affero General Public License as published by
+-- the Free Software Foundation, either version 3 of the License, or
+-- (at your option) any later version.
+--
+-- This program is distributed in the hope that it will be useful,
+-- but WITHOUT ANY WARRANTY; without even the implied warranty of
+-- MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+-- GNU Affero General Public License for more details.
+--
+-- You should have received a copy of the GNU Affero General Public License
+-- along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+{-# LANGUAGE CPP                    #-}
+{-# LANGUAGE Rank2Types             #-}
+{-# LANGUAGE ScopedTypeVariables    #-}
+{-# LANGUAGE FlexibleContexts       #-}
+
+module Language.Haskell.GhcMod.ImportedFrom (importedFrom) where
+
+import Control.Applicative()
+import Control.Exception
+import Control.Monad
+import Control.Monad.Catch
+import Data.ByteString.Internal (w2c)
+import Data.Char (isAlpha)
+import Data.Functor.Identity
+import Data.IORef
+import Data.List
+import Data.List.Split
+import Data.Maybe
+import Data.Typeable()
+import Desugar()
+import Exception (ghandle)
+import FastString
+import GHC
+import GHC.SYB.Utils()
+import HscTypes
+import Language.Haskell.GhcMod
+import Language.Haskell.GhcMod.DynFlags
+import Language.Haskell.GhcMod.FileMapping
+import Language.Haskell.GhcMod.Gap
+import Language.Haskell.GhcMod.GhcPkg
+import Language.Haskell.GhcMod.Logging
+import Language.Haskell.GhcMod.Monad
+import Language.Haskell.GhcMod.Output
+import Language.Haskell.GhcMod.SrcUtils (listifySpans)
+import Outputable
+import System.Directory
+import System.Environment()
+import System.FilePath
+import System.Process
+import System.Process.Streaming
+import TcRnTypes()
+
+import qualified Data.ByteString as B.X
+
+import qualified Data.ByteString as B
+import qualified Data.ByteString.Lazy as BL
+import qualified Data.Map as M
+import qualified Documentation.Haddock as Haddock
+import qualified DynFlags()
+import qualified GhcMonad
+import qualified MonadUtils()
+import qualified Safe
+import qualified SrcLoc
+import qualified Text.Parsec as TP
+
+#if __GLASGOW_HASKELL__ >= 708
```

Comment:
CPP on the GHC version goes in Gap.hs where possible but you really shouldn't use unsafeGlobalDynFlags anyways.